accounting/invoice/models.py

```diff
@@ -20,3 +20,387 @@
+            'total_quantity': aggregated_quantity['total_quantity'],
+            'total_cost': aggregated_cost['total_cost'],
+            'currency': self.hourly_rate.hourly_rate_currency
+        })
+        pdf_name = '{uuid1}-{uuid2}|invoice_{provider}_{client}_{date}.pdf'.format(
+            uuid1=self.uuid,
+            uuid2=uuid.uuid4(),
+            date=self.date.strftime("%Y-%m-%d"),
+            provider=self.provider.user.username,
+            client=self.client.user.username,
+        )
+        pdf_path = os.path.join(settings.INVOICE_PDF_PATH, pdf_name)
+        pdf_configuration = pdfkit.configuration(wkhtmltopdf=settings.HTML_TO_PDF_BINARY_PATH)
+        pdfkit.from_string(invoice, pdf_path, configuration=pdf_configuration, options=self.PDF_OPTIONS)
+        return pdf_path
+
+    def upload_to_google_drive(self, invoice_file):  # NOQA
+        """
+        Upload the invoice in some file format to Google Drive.
+
+        Assumes the invoice has already been turned into a file, i.e. a PDF.
+        Assumes a _very_ specific file structure right now. For example:
+            |- 2016
+            |- 2017
+            |   |- blah
+            |   |- ...
+            |   |- invoices-in
+            |   |   |- 1
+            |   |   |- 2
+            |   |   |- ...
+            |   |   |- 12
+            |   |   |   |- invoice_1.pdf
+            |   |   |   |- ...
+            |   |   |   |- invoice_n.pdf
+            |- 2018
+
+        TODO: This should be a synchronous job on a non-web worker.
+        """
+        if not settings.ENABLE_GOOGLE:
+            return
+
+        # Generate a query we'll use repeatedly for finding the proper folder.
+        query = ("mimeType ='application/vnd.google-apps.folder' and "
+                 "'{parent}' in parents and "
+                 "trashed=false")
+
+        def get_folder_id(root, path=None):
+            """
+            Return the ID of the folder with the last name in `path`, starting traversal from folder given by ID `root`.
+            """
+            if path and isinstance(path, list):
+                folders = drive.ListFile({'q': query.format(parent=root)}).GetList()
+                for folder in folders:
+                    if folder['title'] == path[0]:
+                        root = folder['id']
+                        break
+                return get_folder_id(root, path[1:])
+            return root
+
+        gauth = GoogleAuth()
+        gauth.ServiceAuth()
+        drive = GoogleDrive(gauth)
+        month_folder = get_folder_id(
+            settings.GOOGLE_DRIVE_ROOT,
+            path=[str(self.billing_start_date.year), 'invoices-in', str(self.billing_start_date.month)]
+        )
+        file = drive.CreateFile({
+            'title': invoice_file.split('/')[-1].split('|')[-1],  # TODO: Make this cleaner.
+            'parents': [{'id': month_folder}]
+        })
+        file.SetContentFile(invoice_file)
+        file.Upload()
```

Comment:
This model is pretty big. A part of this logic should probably be moved to dedicated helper functions, for Google Drive and PDF generation, with the model only calling them with parameters?

Reply:
I've thought of that a lot yeah, but it felt faster to just do it right here at first. I'll move them out to mixin classes.

Reply:
I just tried doing this for the PDF generation part, and I see that it's hard to take things out -- there's so much use of `self` for so many different fields, that it makes sense to leave it inside of the Invoice class as is.

Reply:
Hmm - that file is still pretty big, and we're only at the first version. We can leave it as is for now, but we'll probably need to keep this in check and work on it in future evolutions.